src/insertItem.ts

```diff
@@ -0,0 +1,449 @@
+        let part = this.getTemplatePart(template, templatePart);
+        return Json.asArrayValue(part);
+    }
+
+    private getTemplatePart(template: DeploymentTemplate, templatePart: string): Json.Value | undefined {
+        let rootValue = template.topLevelValue;
+        if (!rootValue) {
+            return undefined;
+        }
+        return rootValue.getPropertyValue(templatePart);
+    }
+
+    private async insertParameter(template: DeploymentTemplate, textEditor: vscode.TextEditor): Promise<boolean> {
+        let name = await this.ui.showInputBox({ prompt: "Name of parameter?" });
+        const parameterType = await this.ui.showQuickPick(getItemType(), { placeHolder: 'Type of parameter?' });
+        let parameter: Parameter = {
+            type: parameterType.value
+        };
+        let defaultValue = await this.ui.showInputBox({ prompt: "Default value? Leave empty for no default value.", });
+        if (defaultValue !== '') {
+            parameter.defaultValue = defaultValue;
+        }
+        let description = await this.ui.showInputBox({ prompt: "Description? Leave empty for no description.", });
+        if (description !== '') {
+            parameter.metadata = {
+                description: description
+            };
+        }
+        return await this.insertInObject(template, textEditor, templateKeys.parameters, parameter, name);
+    }
+
+    // tslint:disable-next-line:no-any
+    private async insertInObject(template: DeploymentTemplate, textEditor: vscode.TextEditor, part: string, data: any, name: string): Promise<boolean> {
+        let templatePart = this.getTemplateObjectPart(template, part);
+        if (!templatePart) {
+            let topLevel = template.topLevelValue;
+            if (!topLevel) {
+                vscode.window.showErrorMessage('Invalid ARM template!');
+                return false;
+            }
+            // tslint:disable-next-line:no-any
+            let subPart: any = {};
+            // tslint:disable-next-line:no-unsafe-any
+            subPart[name] = data;
+            await this.insertInObjectInternal(topLevel, textEditor, subPart, part, 1);
+        } else {
+            await this.insertInObjectInternal(templatePart, textEditor, data, name);
+        }
+        return true;
+    }
+
+    // tslint:disable-next-line:no-any
+    private async insertInObjectInternal(templatePart: Json.ObjectValue, textEditor: vscode.TextEditor, data: any, name: string, indentLevel: number = 2): Promise<number> {
+        let isFirstItem = templatePart.properties.length === 0;
+        let startText = isFirstItem ? '' : ',';
+        let index = isFirstItem ? templatePart.span.endIndex : templatePart.properties[templatePart.properties.length - 1].span.afterEndIndex;
+        let tabs = '\t'.repeat(indentLevel - 1);
+        let endText = isFirstItem ? `\r\n${tabs}` : ``;
```

Comment:
I was under the impression that you could use either \r\n or \n and the editor would adjust when inserting (not positive)?

Reply:
The tests still pass after I commented the line
```
text = text.replace(/\r\n/g, '\n');
```
in formatText.
So I removed that line since it was redundant.